### PR TITLE
Implement `PatientPruner`

### PR DESCRIPTION
**Pull request by nzw0301**
## Motivation

Optuna's pruners sometimes prune a trial too aggressively. So, if the objective looks like a noisy loss curve over steps, this aggressive pruning is undesirable because the pruner prunes a promising trial.

For example

- the validation accuracy on neural networks especially, with stochastic procedures; stochastic gradient, dropout, data-augmentation, etc.
- online optimisation

To address this issue, this pull request introduces a wrapper class, `PatientPruner`, of Optuna's pruners based on the discussion in #1447.

`PatientPruner` calls the wrapped pruner the objective function does not improve for `patience` consecutive steps. 

Without `wrapped`, `PatientPruner` can be viewed as the early-stopping mechanism. So the default parameters follow the convention of neural network frameworks such as [`Pytorch ignite`](https://pytorch.org/ignite/master/generated/ignite.handlers.early_stopping.EarlyStopping.html) or [Keras](https://keras.io/api/callbacks/early_stopping/).

Please see #1447 for more details.

## Description of the changes

- [x] Implement PatientPruner
- [x] Documentation
- [x] Add test

### CODE REVIEW

File: optuna/pruners/_patient.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    """
+
+    def __init__(
+        self, wrapped_pruner: Optional[BasePruner], patience: int, min_delta: float = 0.0
+    ) -> None:
+
+        if patience < 0:
+            raise ValueError("patience cannot be negative but got {}.".format(patience))
+
+        self.wrapped_pruner = wrapped_pruner
```
Comment (hvy) [inline optuna/pruners/_patient.py:75]: How about documenting this attribute if made public using the `Attributes:` annotation?

File: optuna/pruners/_patient.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            list(intermediate_values[step] for step in steps_after_patience)
+        )
+
+        direction = study.direction
+        if direction == StudyDirection.MINIMIZE:
+            maybe_prune = np.min(scores_before_patience) - self._min_delta < np.min(
+                scores_after_patience
+            )
+        else:
+            maybe_prune = np.max(scores_before_patience) + self._min_delta > np.min(
```
Comment (hvy) [inline optuna/pruners/_patient.py:107]: Should we take the max here?
```suggestion
            maybe_prune = np.max(scores_before_patience) + self._min_delta > np.max(
```

**Comment by nzw0301** [comment]
Thank you for the feedback! I'll update this pr this week including implements tests.

File: optuna/pruners/_patient.py (hunk 2 above)
Comment (nzw0301) [reply optuna/pruners/_patient.py]: yes! Thank you for suggesting it.

File: optuna/pruners/_patient.py (hunk 1 above)
Comment (nzw0301) [reply optuna/pruners/_patient.py:75]: Indeed, but I think `wrapped_pruned` can be private rather than public because users do not need to access the wrapped pruner.

Comment (hvy) [reply optuna/pruners/_patient.py:75]: Sounds good too. We can always make it public later on  if needed. Let's prefix the property with an underscore in this PR.

**Comment by hvy** [comment]
@HideakiImamura , @himkt could you have a look at this PR? Let me know if you're busy with other tasks.

File: tests/pruners_tests/test_patient.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+from typing import List
+
+import pytest
+
+import optuna
+
+
```
Comment (HideakiImamura) [inline tests/pruners_tests/test_patient.py:7]: How about testing that `PatientPruner` raises `ExperimentalWarning`?

**Review by HideakiImamura — COMMENTED** [review]
The code basically looks good to me. I have several comments. Could you take a look?

File: tests/pruners_tests/test_patient.py (hunk 3 above)
Comment (nzw0301) [reply tests/pruners_tests/test_patient.py:7]: Great!  78e5672

File: optuna/pruners/_patient.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            Wrapped pruner to perform pruning when ``PatientPruner`` allows a
+            trial to be pruned. If it is :obj:`None`, this pruner is equivalent to
+            early-stopping taken the intermediate values in the individual trial.
+        patience:
+            Pruning is disabled until the objective doesn't improve for
+            ``patience`` consecutive steps.
+        min_delta:
+            Tolerance value to check whether or not the objective improves.
+            This value should be non-negative for minimization,
+            and it should be non-positive for maximization.
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply optuna/pruners/_patient.py]: Thank you for your suggestion. Absolutely. it looks more consistent with python's ML ecosystem. As in 884ff98, `min_delta` takes only non-negative value now.

**Review by HideakiImamura — APPROVED** [review]
Thanks for your swift actions! LGTM!

**Comment by nzw0301** [comment]
@himkt any update? :)

File: optuna/pruners/_patient.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        self._patience = patience
+        self._min_delta = min_delta
+
+    def prune(self, study: "optuna.study.Study", trial: "optuna.trial.FrozenTrial") -> bool:
+        step = trial.last_step
+        if step is None:
+            return False
+
+        intermediate_values = trial.intermediate_values
+        steps = np.asarray(list(intermediate_values.keys()))
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply optuna/pruners/_patient.py:88]: but I think `numpy.sort` is faster than built-in `sort`.

**Review by himkt — APPROVED** [review]
Sorry for the late, LGTM.
(Let me put a note that sorting intermediates on each round may cause performance issue)

**Comment by nzw0301** [comment]
@HideakiImamura @himkt  Thank you for reviewing the codes! 

I understand your concern. If we score the best intermediate value somehow instead of searching it every time we call `should_prune`, we can remove the logic. I'm not sure we can do that because optuna allows us to restart training...
